src/Symfony/Component/HttpKernel/EventListener/LocaleAwareListener.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpKernel\EventListener;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpKernel\Event\FinishRequestEvent;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Contracts\Service\LocaleAwareInterface;
+
+/**
+ * Inject request locale to each service that implements LocaleAwareInterface.
+ *
+ * @author Pierre Bobiet <pierrebobiet@gmail.com>
+ */
+class LocaleAwareListener implements EventSubscriberInterface
+{
+    private $localeAwareServices;
+    private $requestStack;
+
+    /**
+     * @param LocaleAwareInterface[] $localeAwareServices
+     */
+    public function __construct(iterable $localeAwareServices, RequestStack $requestStack)
+    {
+        $this->localeAwareServices = $localeAwareServices;
+        $this->requestStack = $requestStack;
+    }
+
+    public function onKernelRequest(GetResponseEvent $event): void
+    {
+        $this->setLocale($event->getRequest());
+    }
+
+    public function onKernelFinishRequest(FinishRequestEvent $event): void
+    {
+        if (null === $parentRequest = $this->requestStack->getParentRequest()) {
+            return;
+        }
+
+        $this->setLocale($parentRequest);
+    }
+
+    public static function getSubscribedEvents()
+    {
+        return array(
+            // must be registered after the Locale listener
+            KernelEvents::REQUEST => array(array('onKernelRequest', 10)),
+            KernelEvents::FINISH_REQUEST => 'onKernelFinishRequest',
+        );
+    }
+
+    private function setLocale(Request $request): void
```

Comment:
As it is implemented right now, a `LocaleAwareListener` cannot be reused in another context than as a listener.
But if we adopt this PR, I'm sure there might be more use cases where we'll need to "sync" the locale.
One example is when handling a message on the receiver side of a message bus.
How do we want to deal with that?